### ISSUE TITLE
EXCEPTION_ACCESS_VIOLATION with SharedArray

### ISSUE THREAD

**Issue by alkorang** [JuliaLang/julia#21631]
There is a function that returns SharedArray.
Spawning in worker 1 works fine, but spawning in other workers raises an exception:
EXCEPTION_ACCESS_VIOLATION
I guess it only happens when I tried to access the elements in SharedArray,
because reading pids after fetch() is not a problem.

example code:
```julia
addprocs(1)
@everywhere function parmul(a, b)
  convert(SharedArray, a * b)
end

N = 2
a = convert(SharedArray, eye(2))
println("Spawn at worker 1:")
c1 = fetch(@spawnat 1 parmul(a, a))
println("c1.pids is $(c1.pids)")
println("c1.pidx is $(c1.pidx)")
println("c1[1, 1] is $(c1[1, 1])")

println("Spawn at worker 2:")
c2 = fetch(@spawnat 2 parmul(a, a))
println("c2.pids is $(c2.pids)")
println("c2.pidx is $(c2.pidx)")
println("c2[1, 1] is $(c2[1, 1])")
```

versioninfo():
```
Julia Version 0.5.1
Commit 6445c82 (2017-03-05 13:25 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, broadwell)
```

output:
```
Spawn at worker 1:
c1.pids is [2]
c1.pidx is 0
c1[1, 1] is 1.0
Spawn at worker 2:
c2.pids is [2]
c2.pidx is 0

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). T
hanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x16f1a6f9 -- _getindex at .\abstractarray.jl:794 [inlined]
getindex at .\abstractarray.jl:760
while loading C:\Users\alkorang\Desktop\Play.jl, in expression starting on line 18
_getindex at .\abstractarray.jl:794 [inlined]
getindex at .\abstractarray.jl:760
jl_call_method_internal at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/s
lave/package_win6_2-x64/build/src\julia_internal.h:210 [inlined]
jl_apply_generic at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/pa
ckage_win6_2-x64/build/src\gf.c:1950
do_call at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win
6_2-x64/build/src\interpreter.c:66
eval at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win6_2
-x64/build/src\interpreter.c:190
do_call at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win
6_2-x64/build/src\interpreter.c:65
eval at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win6_2
-x64/build/src\interpreter.c:190
do_call at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win
6_2-x64/build/src\interpreter.c:65
eval at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win6_2
-x64/build/src\interpreter.c:190
jl_toplevel_eval_flex at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/sla
ve/package_win6_2-x64/build/src\toplevel.c:558
jl_parse_eval_all at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/p
ackage_win6_2-x64/build/src\ast.c:717
jl_load at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_win
6_2-x64/build/src\toplevel.c:596 [inlined]
jl_load_ at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/package_wi
n6_2-x64/build/src\toplevel.c:605
include_from_node1 at .\loading.jl:488
unknown function (ip: 0000000061C614C1)
jl_call_method_internal at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/s
lave/package_win6_2-x64/build/src\julia_internal.h:210 [inlined]
jl_apply_generic at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/pa
ckage_win6_2-x64/build/src\gf.c:1950
process_options at .\client.jl:265
_start at .\client.jl:321
unknown function (ip: 0000000061C8916E)
jl_call_method_internal at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/s
lave/package_win6_2-x64/build/src\julia_internal.h:210 [inlined]
jl_apply_generic at /home/Administrator/buildbot/slave/package_win6_2-x64/build/src/home/Administrator/buildbot/slave/pa
ckage_win6_2-x64/build/src\gf.c:1950
jl_apply at /home/Administrator/buildbot/slave/package_win6_2-x64/build/ui/../src\julia.h:1392 [inlined]
true_main at /home/Administrator/buildbot/slave/package_win6_2-x64/build/ui\repl.c:123
wmain at /home/Administrator/buildbot/slave/package_win6_2-x64/build/ui\repl.c:243
__tmainCRTStartup at /usr/src/debug/mingw64-x86_64-runtime-4.0.6-1/crt\crtexe.c:329
mainCRTStartup at /usr/src/debug/mingw64-x86_64-runtime-4.0.6-1/crt\crtexe.c:212
BaseThreadInitThunk at C:\Windows\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
Allocations: 8629944 (Pool: 8628815; Big: 1129); GC: 11
```

**Comment by amitmurthy** [comment]
Currently the sharedarray segments are mapped onto only workers by default. And mapped onto the the master only if it is created from the master. We should change this to map onto the master in all cases where the requested pids are on the same node as the master. 

A workaround for your case is to explicitly map the shared array on all processes. The changed `parmul` function would look like this:

```
@everywhere function parmul(a, b)
    A = a * b
    S = SharedArray{eltype(A),ndims(A)}(size(A); pids=procs())
    copy!(S, A)
    S
end
```

(On a different note, `parmul` as defined above will not distribute computation across nodes)

**Comment by alkorang** [comment]
I changed `parmul` into two function: `share` and `sharemul`
`share` function simply returns the shared array mapped on all processes.

code:
```julia
addprocs(1)
@everywhere function share(a)
    s = SharedArray(eltype(a), size(a); pids = procs())
    copy!(s, a)
    s
end

@everywhere function sharemul(a, b)
    share(a * b)
end

N = 2
a = convert(SharedArray, eye(N));

println("Spawn at worker 1:")
c1 = fetch(@spawnat 1 sharemul(a, a))
println("c1.pids is $(c1.pids)")
println("c1.pidx is $(c1.pidx)")
println("c1[1, 1] is $(c1[1, 1])")

println("Spawn at worker 2:")
c2 = fetch(@spawnat 2 sharemul(a, a))
println("c2.pids is $(c2.pids)")
println("c2.pidx is $(c2.pidx)")
println("c2[1, 1] is $(c2[1, 1])")
```

output:
```
Spawn at worker 1:
c1.pids is [1,2]
c1.pidx is 1
c1[1, 1] is 1.0
Spawn at worker 2:
c2.pids is [1,2]
c2.pidx is 1
c2[1, 1] is 1.0
```

**Comment by alkorang** [comment]
Fixed in 0.6
